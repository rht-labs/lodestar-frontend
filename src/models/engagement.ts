export interface Engagement {
  name: string;
}
export class Engagement {
  constructor({ name }: Engagement) {
    this.name = name;
  }
  name: string;
}
