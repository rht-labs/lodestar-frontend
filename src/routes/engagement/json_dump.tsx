import React, { useEffect, useState } from 'react';
import { DataCard } from '../../components/engagement_data_cards/data_card';
import AceEditor from 'react-ace';
import 'ace-builds/src-noconflict/mode-json';
import {
  Alert,
  AlertVariant,
  Button,
  Flex,
  FlexItem,
  PageSection,
  PageSectionVariants,
} from '@patternfly/react-core';
import 'ace-builds/src-noconflict/theme-github';
import { Link, useRouteMatch } from 'react-router-dom';
import { Feature } from '../../components/feature/feature';

export interface EngagementJsonDumpProps {
  json: string;
  onSave(json: string): void;
}
export function EngagementJsonDump(props: EngagementJsonDumpProps) {
  const [currentValue, setCurrentValue] = useState<string>(props.json);
  useEffect(() => setCurrentValue(props.json), [props.json]);
  const handleChange = (value: string) => setCurrentValue(value);
  const handleSave = () => props.onSave(currentValue);

  const { url } = useRouteMatch();

  return (
    <DataCard title="Engagement JSON">
      <Alert
        isInline
        title={
          <span>
            Only use this tool if you know what you are doing. If you came here
            by mistake,{' '}
            <Link to={`${url.replace('/json', '')}`}>click here</Link> to return
            to the engagement page.
          </span>
        }
        variant={AlertVariant.danger}
      ></Alert>
      <CodeEditor onChange={handleChange} value={currentValue} />
      <PageSection variant={PageSectionVariants.light}>
        <Flex justifyContent={{ default: 'justifyContentFlexEnd' }}>
          <FlexItem>
            <Feature name="admin">
              <Button onClick={handleSave}>Save</Button>
            </Feature>
          </FlexItem>
        </Flex>
      </PageSection>
    </DataCard>
  );
}

interface CodeEditorProps {
  value: string;
  onChange(value: string): void;
}
const CodeEditor = (props: CodeEditorProps) => {
  return (
    <AceEditor
      width={'100%'}
      mode={'json'}
      value={props.value}
      onChange={props.onChange}
      theme="github"
    />
  );
};
