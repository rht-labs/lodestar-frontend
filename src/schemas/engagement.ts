import faker from 'faker/locale/en_US';
import { LaunchData } from './launch_data';
import { GitCommit } from './git_commit';
import { CreationDetails } from './creation_details';
import { ClusterStatus } from './cluster_status';
export interface Engagement {
  additional_details?: string;
  archive_date: Date;
  commits: GitCommit[];
  customer_contact_email: string;
  customer_contact_name: string;
  customer_name: string;
  description: string;
  end_date: Date;
  engagement_lead_email: string;
  engagement_lead_name: string;
  engagement_users: Array<any>;
  last_update: string;
  location: string;
  mongo_id?: string;
  ocp_cloud_provider_name: string;
  ocp_cloud_provider_region: string;
  ocp_cluster_size: string;
  ocp_persistent_storage_size: string;
  ocp_sub_domain: string;
  ocp_version: string;
  project_id: number;
  project_name: string;
  start_date: Date;
  technical_lead_email: string;
  technical_lead_name: string;
  launch?: LaunchData;
  creation_details: CreationDetails;
  last_update_by_name: string;
  suggested_subdomain?: string;
  status: ClusterStatus;
}
export abstract class Engagement {
  static fromFake(staticData = false): Engagement {
    return {
      additional_details: staticData
        ? 'Additional information here'
        : faker.lorem.paragraphs(2),
      archive_date: staticData ? new Date(2020, 6, 30) : faker.date.recent(),
      commits: [],
      customer_contact_email: staticData
        ? 'bob@doe.com'
        : faker.internet.email(),
      customer_contact_name: staticData
        ? 'Bob Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      customer_name: staticData ? 'NASA' : faker.company.companyName(),
      description: staticData
        ? "It's rocket science"
        : faker.lorem.paragraphs(5),
      end_date: staticData ? new Date(2020, 6, 1) : faker.date.future(),
      engagement_users: staticData ? [] : [],
      engagement_lead_email: staticData
        ? 'alice@doe.com'
        : faker.internet.email(),
      engagement_lead_name: staticData
        ? 'Alice Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      last_update: staticData ? '2020-01-01' : new Date().toString(),
      location: staticData
        ? 'Nashville, TN'
        : `${faker.address.city()}, ${faker.address.stateAbbr()}`,
      mongo_id: '1',
      ocp_cloud_provider_name: staticData ? 'AWS' : 'AWS',
      ocp_cloud_provider_region: staticData
        ? 'N. Virginia'
        : faker.lorem.word(),
      ocp_cluster_size: staticData ? 'Large' : 'Large',
      ocp_persistent_storage_size: staticData
        ? '100G'
        : `${faker.random.number()}G`,
      ocp_sub_domain: staticData ? 'ordoabchao' : faker.lorem.word(),
      ocp_version: staticData ? '4.4' : faker.random.number().toString(),
      project_id: staticData ? 1 : faker.random.number(),
      project_name: staticData ? 'Boots on the Moon' : faker.company.bsNoun(),
      start_date: staticData ? new Date(2020, 5, 1) : faker.date.recent(),
      technical_lead_email: staticData ? 'eve@doe.com' : faker.internet.email(),
      technical_lead_name: staticData
        ? 'Eve Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      creation_details: staticData
        ? {
            created_by_email: 'dwasinger@redhat.com',
            created_by_user: 'dwasinge',
            created_on: new Date(2020, 1, 1),
          }
        : {
            created_by_email: faker.internet.email(),
            created_by_user: `${faker.name.firstName()} ${faker.name.lastName()}`,
            created_on: faker.date.recent(),
          },
      last_update_by_name: staticData
        ? 'James Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      launch: staticData
        ? null
        : faker.random.boolean()
        ? {
            launched_by: faker.name.firstName(),
            launched_date_time: faker.date.recent(),
          }
        : null,
      status: ClusterStatus.fromFake(staticData),
    };
  }
}

export enum EngagementStatus {
  active = 'active',
  past = 'past',
  upcoming = 'upcoming',
}

export const getEngagementStatus = (
  engagement: Engagement
): EngagementStatus => {
  if (!engagement) {
    return null;
  }
  const Today = new Date();
  const { launch, end_date } = engagement;
  const hasLaunched = !!launch?.launched_date_time;
  if (hasLaunched && end_date >= Today) {
    return EngagementStatus.active;
  } else if (hasLaunched && end_date < Today) {
    return EngagementStatus.past;
  } else {
    return EngagementStatus.upcoming;
  }
};
