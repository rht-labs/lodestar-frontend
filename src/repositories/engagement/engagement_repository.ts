import { Engagement } from '../../models/engagement';

export abstract class EngagementRepository {
  abstract async fetchEngagements(): Promise<Engagement[]>;
}
