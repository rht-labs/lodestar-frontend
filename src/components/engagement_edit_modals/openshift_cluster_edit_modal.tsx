import React from 'react';
import { Engagement } from '../../schemas/engagement_schema';
import { Modal, ModalVariant, Button, Form } from '@patternfly/react-core';
import { useModalVisibility } from '../../context/edit_modal_visibility_context/edit_modal_visibility_hook';
import { EditModalTemplate } from '../../layout/edit_modal_template';
import {
  EngagementFormConfig,
  EngagementFormOption,
} from '../../schemas/engagement_config';
import { CloudProviderFormField } from '../engagement_form_fields/cloud_provider';
import { CloudProviderRegionFormField } from '../engagement_form_fields/cloud_provider_region';
import { OpenShiftVersionFormField } from '../engagement_form_fields/oc_version';
import { SubdomainFormField } from '../engagement_form_fields/subdomain';
import { PersistentStorageFormField } from '../engagement_form_fields/persistent_storage';
import { ClusterSizeFormField } from '../engagement_form_fields/cluster_size';
import { AdditionalDetailsFormField } from '../engagement_form_fields/additional_details';
export interface OpenShiftClusterEditModalProps {
  onChange: (fieldName: string, value: any) => void;
  formOptions: EngagementFormConfig;
  engagement: Engagement;
  isOpen: boolean;
  onSave: (engagement: Engagement) => void;
}

export function OpenShiftClusterEditModal({
  engagement,
  formOptions,
  onChange,
  isOpen,
  onSave: propsOnSave,
}: OpenShiftClusterEditModalProps) {
  const { requestClose } = useModalVisibility();
  const availableProviders = getAvailableProviders(engagement, formOptions);
  const provider = formOptions?.cloud_options?.providers?.options?.find(
    p => p.value === engagement?.ocp_cloud_provider_name
  );
  const availableProviderRegionOptions = getAvailableRegionOptions(
    provider,
    engagement
  );

  const onSave = () => {
    propsOnSave(engagement);
    requestClose();
  };
  return (
    <Modal
      variant={ModalVariant.large}
      isOpen={isOpen}
      onClose={requestClose}
      title="Hosting Environment"
    >
      <EditModalTemplate
        actions={
          <div>
            <Button onClick={onSave} data-cy={'hosting_env_save'}>Save</Button>
          </div>
        }
      >
        <Form isHorizontal>
          <CloudProviderFormField
            onChange={onChange}
            availableProviders={availableProviders}
            engagement={engagement}
          />
          <CloudProviderRegionFormField
            onChange={onChange}
            availableProviderRegionOptions={availableProviderRegionOptions}
            engagement={engagement}
          />
          <OpenShiftVersionFormField
            onChange={onChange}
            formOptions={formOptions}
            engagement={engagement}
          />
          <SubdomainFormField
            onChange={onChange}
            formOptions={formOptions}
            engagement={engagement}
          />
          <PersistentStorageFormField
            onChange={onChange}
            formOptions={formOptions}
            engagement={engagement}
          />
          <ClusterSizeFormField
            onChange={onChange}
            formOptions={formOptions}
            engagement={engagement}
          />
          <AdditionalDetailsFormField
            onChange={onChange}
            engagement={engagement}
          />
        </Form>
      </EditModalTemplate>
    </Modal>
  );
}

function getAvailableProviders(
  engagement: Engagement,
  formOptions: EngagementFormConfig
) {
  const availableProviders =
    formOptions?.cloud_options?.providers?.options ?? [];
  if (
    engagement?.ocp_cloud_provider_name &&
    !availableProviders.find(
      option =>
        option.value && option.value === engagement?.ocp_cloud_provider_name
    )
  ) {
    availableProviders.push({
      value: engagement?.ocp_cloud_provider_name,
      label: engagement?.ocp_cloud_provider_name,
    });
  }
  return availableProviders;
}

function getAvailableRegionOptions(
  provider: EngagementFormOption,
  engagement: Engagement
) {
  const availableProviderRegionOptions = provider?.options ?? [];

  if (
    availableProviderRegionOptions?.length === 0 &&
    engagement?.ocp_cloud_provider_region
  ) {
    availableProviderRegionOptions.push({
      value: engagement?.ocp_cloud_provider_region,
      label: engagement?.ocp_cloud_provider_region,
    });
  }
  return availableProviderRegionOptions;
}
