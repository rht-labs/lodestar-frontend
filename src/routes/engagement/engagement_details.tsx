import React, { useEffect } from 'react';
import { Engagement } from '../../schemas/engagement';
import { useEngagement } from '../../context/engagement_context/engagement_hook';
import { Route, useParams, Switch, useRouteMatch } from 'react-router';
import { getValidatorsFromEngagementFormConfig } from '../../common/config_validator_adapter';
import {
  ValidationProvider,
  FormValidator,
} from '../../context/validation_context/validation_context';
import { EngagementDetailsViewTemplate } from '../../layout/engagement_details_view';
import { EngagementOverview } from './overview';
import { EngagementJsonDump } from './json_dump';
import { EngagementJsonSerializer } from '../../serializers/engagement/engagement_json_serializer';

export interface EngagementViewProps {
  currentEngagement?: Engagement;
}

export interface EngagementDetailViewProps {
  error: Error;
  currentEngagement: Engagement;
  createEngagementPoll: (
    engagement: Engagement
  ) => Promise<{ cancel: () => void }>;
  getEngagement: (
    customer_name: string,
    project_name: string
  ) => Promise<Engagement>;
}
export const validatorsWithDateValidators = (
  validators: FormValidator,
  dates: Pick<Engagement, 'start_date' | 'end_date'>
): FormValidator => {
  return {
    ...validators,
    end_date: [
      ...(validators['end_date'] ?? []),
      endDate => {
        const result =
          dates?.start_date > endDate
            ? 'End date must be after the start date'
            : null;
        return result;
      },
    ],
    archive_date: [
      ...(validators['archive_date'] ?? []),
      archiveDate => {
        const result =
          archiveDate < dates?.end_date
            ? 'Archive date must be after the end date'
            : null;
        return result;
      },
    ],
  };
};

export const EngagementDetailView = () => {
  const { project_name, customer_name } = useParams<{
    customer_name: string;
    project_name: string;
  }>();

  const {
    createEngagementPoll,
    saveEngagement,
    currentEngagement,
    currentChanges,
    engagementFormConfig,
  } = useEngagement({ projectName: project_name, customerName: customer_name });

  useEffect(() => {
    let engagementPoll;
    if (currentEngagement?.project_name && currentEngagement?.customer_name) {
      engagementPoll = createEngagementPoll(currentEngagement);
    }
    return async () => {
      (await engagementPoll)?.cancel?.();
    };
  }, [currentEngagement, createEngagementPoll]);

  const validators = getValidatorsFromEngagementFormConfig(
    engagementFormConfig
  );
  const { url } = useRouteMatch();
  const serializer = new EngagementJsonSerializer();
  const {
    commits,
    uuid,
    project_id,
    mongo_id,
    creation_details,
    status,
    launch,
    ...editableFields
  } = currentEngagement ?? {};
  const { end_date, start_date } = currentChanges;

  return (
    <ValidationProvider
      validators={validatorsWithDateValidators(validators, {
        end_date,
        start_date,
      })}
    >
      <EngagementDetailsViewTemplate
        engagement={currentEngagement}
        onSave={saveEngagement}
      >
        <Switch>
          <Route path={`${url}/json`}>
            <EngagementJsonDump
              json={JSON.stringify(
                serializer.serialize((editableFields ?? {}) as Engagement),
                null,
                2
              )}
              onSave={value => {
                saveEngagement(serializer.deserialize(JSON.parse(value)));
              }}
            />
          </Route>
          <Route path={`${url}/`}>
            <EngagementOverview />
          </Route>
        </Switch>
      </EngagementDetailsViewTemplate>
    </ValidationProvider>
  );
};
