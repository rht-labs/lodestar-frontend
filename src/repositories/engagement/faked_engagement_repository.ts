import { EngagementRepository } from './engagement_repository';
import { Engagement } from '../../models/engagement';
import faker from 'faker';

export class FakedEngagementRepository extends EngagementRepository {
  async fetchEngagements(): Promise<Engagement[]> {
    return new Array(8)
      .fill(null)
      .map(() => new Engagement({ name: faker.company.companyName() }));
  }
}
