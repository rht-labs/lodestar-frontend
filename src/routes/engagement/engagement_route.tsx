import React from 'react';
import { useAnalytics } from '../../context/analytics_context/analytics_context';
import { useSession } from '../../context/auth_context/auth_context';
import { useEngagementFormConfig } from '../../context/engagement_config_context/engagement_config_hook';
import { EngagementProvider } from '../../context/engagement_context/engagement_context';
import { useFeedback } from '../../context/feedback_context/feedback_context';
import { useServiceProviders } from '../../context/service_provider_context/service_provider_context';
import ScrollToTop from '../scroll_to_top';
import { EngagementDetailView } from './engagement_details';

export const EngagementRoute = () => {
  const { engagementService, categoryService } = useServiceProviders();
  const feedback = useFeedback();
  const analyticsContext = useAnalytics();
  const authContext = useSession();
  const { engagementFormConfig } = useEngagementFormConfig(engagementService);
  return (
    <EngagementProvider
      authContext={authContext}
      feedbackContext={feedback}
      analyticsContext={analyticsContext}
      categoryService={categoryService}
      engagementService={engagementService}
      engagementFormConfig={engagementFormConfig}
    >
      <ScrollToTop>
        <EngagementDetailView />
      </ScrollToTop>
    </EngagementProvider>
  );
};
