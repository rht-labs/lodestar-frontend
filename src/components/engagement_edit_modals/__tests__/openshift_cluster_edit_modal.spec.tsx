import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { OpenShiftClusterEditModal } from '../openshift_cluster_edit_modal';
import { Engagement } from '../../../schemas/engagement';

describe('Point of Contact edit modal', () => {
  test('matches snapshot', () => {
    expect(
      render(
        <OpenShiftClusterEditModal
          onSave={() => {}}
          formOptions={{}}
          isOpen={true}
          engagement={Engagement.fromFake(true)}
          onChange={() => {}}
        />
      )
    ).toMatchSnapshot();
  });

  test('When clicking the save button, the onSave method is called', async () => {
    const onSave = jest.fn();
    const { getByTestId } = render(
      <OpenShiftClusterEditModal
        onSave={onSave}
        engagement={Engagement.fromFake(true)}
        formOptions={{}}
        isOpen={true}
        onChange={() => {}}
      />
    );
    await fireEvent.click(getByTestId('oc-edit-save'));
    expect(onSave).toHaveBeenCalled();
  });
});
