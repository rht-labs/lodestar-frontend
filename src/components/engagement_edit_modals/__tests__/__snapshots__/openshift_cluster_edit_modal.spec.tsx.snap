// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Point of Contact edit modal matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="pf-c-backdrop__open"
  >
    <div
      aria-hidden="true"
    />
    <div>
      <div
        class="pf-c-backdrop"
      >
        <div
          class="pf-l-bullseye"
        >
          <div
            aria-describedby="pf-modal-part-2"
            aria-labelledby="pf-modal-part-1"
            aria-modal="true"
            class="pf-c-modal-box pf-m-lg"
            data-ouia-component-id="0"
            data-ouia-component-type="PF4/ModalContent"
            data-ouia-safe="true"
            id="pf-modal-part-0"
            role="dialog"
          >
            <button
              aria-disabled="false"
              aria-label="Close"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  transform=""
                />
              </svg>
            </button>
            <header
              class="pf-c-modal-box__header"
            >
              <h1
                class="pf-c-modal-box__title pf-c-modal-box__title"
                id="pf-modal-part-1"
              >
                Hosting Environment
              </h1>
            </header>
            <div
              class="pf-c-modal-box__body"
              id="pf-modal-part-2"
            >
              <section
                class="pf-c-page__main-section"
              >
                <form
                  class="pf-c-form pf-m-horizontal"
                  novalidate=""
                >
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="Hosting Platform"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Hosting Type
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <select
                        aria-invalid="false"
                        class="pf-c-form-control"
                        data-ouia-component-id="2"
                        data-ouia-component-type="PF4/FormSelect"
                        data-ouia-safe="true"
                        disabled=""
                        required=""
                      >
                        <option
                          class=""
                          value=""
                        >
                          OpenShift Container Platform
                        </option>
                      </select>
                      
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="cloud-provider"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Cloud Provider
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           
                          *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <select
                        aria-invalid="false"
                        aria-label="Cloud Provider"
                        class="pf-c-form-control"
                        data-ouia-component-id="3"
                        data-ouia-component-type="PF4/FormSelect"
                        data-ouia-safe="true"
                        data-testid="cloud-provider-select"
                        disabled=""
                        id="cloud_provider_dropdown"
                        required=""
                      >
                        <option
                          class=""
                          value=""
                        >
                          Select a provider
                        </option>
                        <option
                          class=""
                          data-cy="AWS"
                          value="AWS"
                        >
                          AWS
                        </option>
                      </select>
                      
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="cloud-provider-region"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Provider Region
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           
                          *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <select
                        aria-invalid="false"
                        aria-label="Cloud provider region"
                        class="pf-c-form-control"
                        data-ouia-component-id="4"
                        data-ouia-component-type="PF4/FormSelect"
                        data-ouia-safe="true"
                        data-testid="provider-region-select"
                        disabled=""
                        id="cloud_provider_region_dropdown"
                      >
                        <option
                          class=""
                          value=""
                        >
                          Select a region
                        </option>
                        <option
                          class=""
                          data-cy="N. Virginia"
                          disabled=""
                          value="N. Virginia"
                        >
                          N. Virginia
                        </option>
                      </select>
                      
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="openshift-provider"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          OpenShift Version
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           
                          *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <select
                        aria-invalid="false"
                        aria-label="OpenShift Version"
                        class="pf-c-form-control"
                        data-ouia-component-id="5"
                        data-ouia-component-type="PF4/FormSelect"
                        data-ouia-safe="true"
                        data-testid="oc-version-select"
                        disabled=""
                        id="oc_version_dropdown"
                      >
                        <option
                          class=""
                          value=""
                        >
                          Select OpenShift version
                        </option>
                      </select>
                      
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="subdomain"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Desired Subdomain
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           
                          *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <input
                        aria-invalid="false"
                        class="pf-c-form-control"
                        data-cy="desired_subdomain_input"
                        data-testid="subdomain-input"
                        disabled=""
                        id="ocp_sub_domain"
                        name="ocp_sub_domain"
                        required=""
                        type="text"
                        value="ordoabchao"
                      />
                      <div>
                        Applications will live at: 
                        <strong>
                          ordoabchao.na-1.rht-labs.com
                        </strong>
                      </div>
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="persistent-storage-needs"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Persistent Storage Needs
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           
                          *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <select
                        aria-invalid="false"
                        aria-label="Persistent Storage Needs"
                        class="pf-c-form-control"
                        data-ouia-component-id="6"
                        data-ouia-component-type="PF4/FormSelect"
                        data-ouia-safe="true"
                        data-testid="persistent-storage-select"
                        disabled=""
                        id="persistent_storage_dropdown"
                      >
                        <option
                          class=""
                          value=""
                        />
                      </select>
                      
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="cluster-size"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Cluster Size
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           
                          *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <select
                        aria-invalid="false"
                        aria-label="Cluster Size"
                        class="pf-c-form-control"
                        data-ouia-component-id="7"
                        data-ouia-component-type="PF4/FormSelect"
                        data-ouia-safe="true"
                        data-testid="cluster-size-select"
                        disabled=""
                        id="cluster_size_dropdown"
                        required=""
                      >
                        <option
                          class=""
                          value=""
                        >
                          Select cluster size
                        </option>
                      </select>
                      
                    </div>
                  </div>
                  <div
                    class="pf-c-form__group"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="additional_details"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Additional Information
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <textarea
                        aria-invalid="false"
                        aria-label="Additional provisioning information"
                        class="pf-c-form-control pf-m-resize-vertical"
                        data-testid="additional-details-text-area"
                        disabled=""
                        id="additional_details"
                        name="additional_details"
                        placeholder="Provide additional details"
                      >
                        Additional information here
                      </textarea>
                      
                    </div>
                  </div>
                </form>
              </section>
              <section
                class="pf-c-page__main-section"
              >
                <div
                  style="margin-top: 1rem;"
                >
                  <div
                    class="pf-l-flex pf-m-justify-content-flex-end"
                  >
                    <div
                      class=""
                    >
                      <div>
                        <button
                          aria-disabled="false"
                          class="pf-c-button pf-m-primary"
                          data-cy="hosting_env_save"
                          data-ouia-component-id="8"
                          data-ouia-component-type="PF4/Button"
                          data-ouia-safe="true"
                          data-testid="oc-edit-save"
                          type="button"
                        >
                          Save
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </section>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
